dm/syncer/data_validator.go

```diff
@@ -198,3 +399,158 @@ func (v *DataValidator) Stage() pb.Stage {
+		return nil
+	}
+
+	table := &validateTableInfo{
+		Source:     sourceTable,
+		Info:       tableInfo,
+		PrimaryKey: primaryIdx,
+		Target:     targetTable,
+	}
+
+	for _, cols := range ev.SkippedColumns {
+		if len(cols) > 0 {
+			err := errors.Errorf("unexpected skipped columns for table %s", sourceTable.String())
+			return err
+		}
+	}
+
+	changeType := getRowChangeType(header.EventType)
+	v.changeEventCount[changeType].Inc()
+
+	columnMap := make(map[string]*model.ColumnInfo)
+	for _, col := range tableInfo.Columns {
+		columnMap[col.Name.O] = col
+	}
+	pk := table.PrimaryKey
+	pkIndices := make([]int, len(pk.Columns))
+	for i, col := range pk.Columns {
+		pkIndices[i] = columnMap[col.Name.O].Offset
+	}
+	table.pkIndices = pkIndices
+
+	step := 1
+	if changeType == rowUpdated {
+		step = 2
+	}
+	for i := 0; i < len(ev.Rows); i += step {
+		row := ev.Rows[i]
+		pkValue := make([]string, len(pk.Columns))
+		for _, idx := range pkIndices {
+			pkValue[idx] = string(genColData(row[idx]))
+		}
+		key := genRowKey(pkValue)
+
+		if changeType == rowUpdated {
+			afterRowChangeType := changeType
+			afterRow := ev.Rows[i+1]
+			afterPkValue := make([]string, len(pk.Columns))
+			for _, idx := range pkIndices {
+				afterPkValue[idx] = string(genColData(afterRow[idx]))
+			}
+			afterKey := genRowKey(afterPkValue)
+			if afterKey != key {
+				// convert to delete and insert
+				v.dispatchRowChange(key, &rowChange{
+					table:      table,
+					key:        key,
+					pkValues:   pkValue,
+					data:       row,
+					tp:         rowDeleted,
+					lastMeetTS: int64(header.Timestamp),
+				})
+				afterRowChangeType = rowInsert
+			}
+			v.dispatchRowChange(afterKey, &rowChange{
+				table:      table,
+				key:        afterKey,
+				pkValues:   afterPkValue,
+				data:       afterRow,
+				tp:         afterRowChangeType,
+				lastMeetTS: int64(header.Timestamp),
+			})
+		} else {
+			v.dispatchRowChange(key, &rowChange{
+				table:      table,
+				key:        key,
+				pkValues:   pkValue,
+				data:       row,
+				tp:         changeType,
+				lastMeetTS: int64(header.Timestamp),
+			})
+		}
+	}
+	return nil
+}
+
+// getRowChangeType should be called only when the event type is RowsEvent.
+func getRowChangeType(t replication.EventType) rowChangeType {
+	switch t {
+	case replication.WRITE_ROWS_EVENTv0, replication.WRITE_ROWS_EVENTv1, replication.WRITE_ROWS_EVENTv2:
+		return rowInsert
+	case replication.UPDATE_ROWS_EVENTv0, replication.UPDATE_ROWS_EVENTv1, replication.UPDATE_ROWS_EVENTv2:
+		return rowUpdated
+	default:
+		// replication.DELETE_ROWS_EVENTv0, replication.DELETE_ROWS_EVENTv1, replication.DELETE_ROWS_EVENTv2:
+		return rowDeleted
+	}
+}
+
+func genRowKey(pkValues []string) string {
+	return strings.Join(pkValues, "-")
+}
+
+func genColData(v interface{}) []byte {
+	switch dv := v.(type) {
+	case []byte:
+		return dv
+	case string:
+		return []byte(dv)
+	case decimal.Decimal:
```

Comment:
we didn't set UseDecimal in BinlogSyncerConfig, so in fact will not meet this type